src/Features/LanguageServer/Protocol/Handler/Diagnostics/AbstractWorkspacePullDiagnosticsHandler.cs

```diff
@@ -0,0 +1,252 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.Host;
+using Microsoft.CodeAnalysis.Options;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Microsoft.CodeAnalysis.SolutionCrawler;
+using Microsoft.CodeAnalysis.TaskList;
+using Microsoft.VisualStudio.LanguageServer.Protocol;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.LanguageServer.Handler.Diagnostics;
+internal abstract class AbstractWorkspacePullDiagnosticsHandler<TDiagnosticsParams, TReport, TReturn> : AbstractPullDiagnosticHandler<TDiagnosticsParams, TReport, TReturn>, IDisposable
+    where TDiagnosticsParams : IPartialResultParams<TReport>
+{
+    private readonly LspWorkspaceRegistrationService _workspaceRegistrationService;
+    private readonly LspWorkspaceManager _workspaceManager;
+
+    /// <summary>
+    /// Flag that represents whether the LSP view of the world has changed.
+    /// It is totally fine for this to somewhat over-report changes
+    /// as it is an optimization used to delay closing workspace diagnostic requests
+    /// until something has changed.
+    /// </summary>
+    private int _lspChanged = 0;
+
+    protected AbstractWorkspacePullDiagnosticsHandler(
+        LspWorkspaceManager workspaceManager,
+        LspWorkspaceRegistrationService registrationService,
+        IDiagnosticAnalyzerService diagnosticAnalyzerService,
+        IDiagnosticsRefresher diagnosticRefresher,
+        IGlobalOptionService globalOptions) : base(diagnosticAnalyzerService, diagnosticRefresher, globalOptions)
+    {
+        _workspaceManager = workspaceManager;
+        _workspaceRegistrationService = registrationService;
+
+        _workspaceRegistrationService.LspSolutionChanged += OnLspSolutionChanged;
+        _workspaceManager.LspTextChanged += OnLspTextChanged;
+    }
+
+    public void Dispose()
+    {
+        _workspaceManager.LspTextChanged -= OnLspTextChanged;
+        _workspaceRegistrationService.LspSolutionChanged -= OnLspSolutionChanged;
+    }
+
+    protected override async ValueTask<ImmutableArray<IDiagnosticSource>> GetOrderedDiagnosticSourcesAsync(TDiagnosticsParams diagnosticsParams, RequestContext context, CancellationToken cancellationToken)
```

Comment:
All of this (and related methods) was extracted from WorkspacePullDiagnosticsHandler unchanged.